common/flatpak-installation.c

```diff
@@ -1024,76 +1052,114 @@ flatpak_installation_list_installed_refs_for_update (FlatpakInstallation *self,
-      g_hash_table_insert (remote_states, (char *)remote, g_steal_pointer (&state));
     }
 
-  updates = g_ptr_array_new_with_free_func (g_object_unref);
+  related_to_refs = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
 
-  if (flatpak_dir_list_refs (dir, "app", &refs_app, cancellable, error))
+  g_signal_connect (transaction, "ready", G_CALLBACK (transaction_ready), &related_to_refs);
+
+  flatpak_transaction_run (transaction, cancellable, &local_error);
+  if (!g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_ABORTED))
     {
-     for (int i = 0; refs_app[i] != NULL; i++)
-       {
-        const char *ref = refs_app[i];
-        if (ref_check_for_update (dir, ref, remote_states, cancellable))
-          {
-            g_debug ("%s: Installed ref %s needs update", G_STRFUNC, ref);
-            FlatpakInstalledRef *installed_ref = get_ref (dir, ref, cancellable, NULL);
-            if (installed_ref)
-              g_ptr_array_add (updates, g_object_ref (installed_ref));
-          }
-       }
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return NULL;
     }
-
-  if (flatpak_dir_list_refs (dir, "runtime", &refs_runtime, cancellable, error))
+  g_clear_error (&local_error);
+
+  installed_refs_for_update = g_ptr_array_new_with_free_func (g_object_unref);
+  installed_refs_used = g_new0 (gboolean, installed_refs->len);
+
+  /* For each ref that would be affected by the transaction, if it is
+   * installed, add it to the list to be returned and otherwise add the ref
+   * that caused it be added, "related_to_ref". We need to cover all of the
+   * following cases:
+   * 1. For an app or runtime that has an update available, add it to the list.
+   * 2. For an app or extension that has a missing runtime, add the
+   * app/extension to the list.
+   * 3. For an app that's missing a "should-download" related ref, add the app
+   * to the list.
+   */
+  GLNX_HASH_TABLE_FOREACH_KV (related_to_refs, char *, ref, char *, related_to_ref)
     {
-     for (int i = 0; refs_runtime[i] != NULL; i++)
-       {
-        const char *ref = refs_runtime[i];
-        if (ref_check_for_update (dir, ref, remote_states, cancellable))
-          {
-            g_debug ("%s: Installed ref %s needs update", G_STRFUNC, ref);
-            FlatpakInstalledRef *installed_ref = get_ref (dir, ref, cancellable, NULL);
-            if (installed_ref)
-              g_ptr_array_add (updates, g_object_ref (installed_ref));
-          }
-       }
+      guint index;
+
+      /* Here we use the existing installed_refs array instead of get_ref()
+       * since staying in memory should be more efficient than disk I/O
+       * despite that the search is O(n).
+       */
+      if (g_ptr_array_find_with_equal_func (installed_refs, ref,
+                                            (GEqualFunc)installed_ref_equal_func,
+                                            &index))
+        {
+          if (!installed_refs_used[index])
```

Comment:
You could use hashtables for `installed_refs` for faster lookup than the array search, as well as for `installed_refs_for_update` to avoid needing the extra check for if you already added it.

Reply:
Yeah, although in the case of `installed_refs_for_update` we then need to translate back into a pointer array to avoid breaking API.

Reply:
Never mind, we can use a pointer array and hash table in parallel, since the hash table will use strings and the pointer array `FlatpakInstalledRef`.